### PR TITLE
Voxels structure, transforms, conversion func and ModelNet10/40

### PR DESCRIPTION
**Pull request by nirmal-suthar**
Implementation of VoxelGrid (voxels) and Interconversion functions and ModelNet10/40.

- [x] VoxelGrid representation
- [x] Updating ModelNet10/40 (building from official zip)
- [x] Improving forward/back pass of chamfer distance on GPU/CPU. 
- [x] Interconversion function for converting to different structures.
- [x] `visualize` function for voxels 
- [x] Transforms for conversions
- [x] Add GPU support using CUDA.jl 
- [x] Add relevant test

### CODE REVIEW

File: src/conversions.jl [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+    voxels = S{T,4}(undef,res,res,res,_B)
+    # print(typeof(verts_list))
+    for (i,v,f) in zip(1:_B,verts_list, faces_list)
+        voxels[:,:,:,i] = _voxelize(cpu(v),f,res)
+    end
+    return voxels
+end
+
+function _voxelize(
+    v::Array{<:AbstractFloat,2},
```
Comment (DhairyaLGandhi) [inline src/conversions.jl:90]: Does it need to be abstract float here, what about real?

File: src/conversions.jl [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+                        dims = 2,
+                    )
+                end
+            end
+        end
+
+        v1, v2, v3 = new_verts
+    end
+
+    voxels = zeros(resolution, resolution, resolution)
```
Comment (DhairyaLGandhi) [inline src/conversions.jl:151]: Might be an issue for cuarrays

File: src/conversions.jl (hunk 1 above)
Comment (nirmal-suthar) [reply src/conversions.jl:90]: verts in TriMesh is restricted to AbstractFloat, therefore I only included the same here, although changing to Real won't be any problem

File: src/conversions.jl (hunk 2 above)
Comment (nirmal-suthar) [reply src/conversions.jl:151]: This function only expects array and later we convert to CuArray according to the original input. The problem with cuarray in this function was that logical indexing is not supported by it, and some lower function also uses other libraries like Meshing.jl which again expects Array.

Comment (DhairyaLGandhi) [reply src/conversions.jl:151]: Hmm, is there advantage in having meshing.jl have cuda compatible kernels?

Comment (nirmal-suthar) [reply src/conversions.jl:151]: Meshing.jl is fast even on cpu. For average size TriMesh (2000 verts) converting to voxels takes ~1ms.

**Comment by nirmal-suthar** [comment]
@avik-pal @DhairyaLGandhi the error on GPU ci is a bit weird, error is related to unzip command, and it is running fine with GitHub ci but GPU ci is throwing error that directory not found.

**Comment by nirmal-suthar** [comment]
Other than that, I believe this pr is ready to be merged. For doc changes, opening another pr would be better.

**Comment by avik-pal** [comment]
I am not sure what is going on there. On a sidenote maybe we should not be redistributing datasets like that unless their LICENSE permits us to do so.

File: src/conversions.jl [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+        side_1 = sum((v1 .- v2) .^ 2, dims = 1)
+        side_2 = sum((v2 .- v3) .^ 2, dims = 1)
+        side_3 = sum((v3 .- v1) .^ 2, dims = 1)
+        sides = cat(side_1, side_2, side_3, dims = 1)
+        sides = reshape(maximum(sides, dims = 1), :)
+
+        keep = sides .> smallest_side
+        if !(any(keep))
+            break
+        end
```
Comment (avik-pal) [inline src/conversions.jl:115]: Minor style nit
```suggestion
        !(any(keep)) && break
```

File: src/conversions.jl (hunk 2 above)
Comment (avik-pal) [reply src/conversions.jl:151]: I am okay with having only cpu versions for now. These functions are generally used once for the lifetime of a project and are rarely a bottleneck. However, I believe opening an issue (or ask on slack) in Meshing.jl to ask if there are plans to include cuda kernels in the future would be worthwhile.

**Review by avik-pal — APPROVED** [review]

**Comment by nirmal-suthar** [comment]
> I am not sure what is going on there. On a sidenote maybe we should not be redistributing datasets like that unless their LICENSE permits us to do so.

I did some searching for such LICENSE but could not find any info on this :(

**Comment by avik-pal** [comment]
> > I am not sure what is going on there. On a sidenote maybe we should not be redistributing datasets like that unless their LICENSE permits us to do so.
> 
> I did some searching for such LICENSE but could not find any info on this :(

Yeah best not to redistribute it.

**Comment by nirmal-suthar** [comment]
Sure, will remove the option to autodownload

**Comment by avik-pal** [comment]
Maybe disable the tests for dataset on GPU? Unless ofcourse they are testing anything GPU specific

**Comment by nirmal-suthar** [comment]
> Maybe disable the tests for dataset on GPU? Unless ofcourse they are testing anything GPU specific

Datasets tests are not GPU specific, so will make a temp fix to block this test on gpu.

**Pull request merged**